core/src/avm2/globals/array.rs

```diff
@@ -0,0 +1,1353 @@
+                let item = resolve_array_hole(activation, this, i, item)?;
+
+                accum.push(
+                    conv(item, activation)?
+                        .coerce_to_string(activation)?
+                        .to_string(),
+                );
+            }
+
+            return Ok(AvmString::new(
+                activation.context.gc_context,
+                accum.join(&string_separator),
+            )
+            .into());
+        }
+    }
+
+    Ok(Value::Undefined)
+}
+
+/// Implements `Array.join`
+pub fn join<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    join_inner(activation, this, args, |v, _act| Ok(v))
+}
+
+/// Implements `Array.toString`
+pub fn to_string<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    join_inner(activation, this, &[",".into()], |v, _act| Ok(v))
+}
+
+/// Implements `Array.toLocaleString`
+pub fn to_locale_string<'gc>(
+    act: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    join_inner(act, this, &[",".into()], |v, activation| {
+        let mut o = v.coerce_to_object(activation)?;
+
+        let tls = o.get_property(
+            o,
+            &QName::new(Namespace::public_namespace(), "toLocaleString"),
+            activation,
+        )?;
+
+        tls.coerce_to_object(activation)?
+            .call(Some(o), &[], activation, o.proto())
+    })
+}
+
+/// Implements `Array.valueOf`
+pub fn value_of<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    join_inner(activation, this, &[",".into()], |v, _act| Ok(v))
+}
+
+/// An iterator that allows iterating over the contents of an array whilst also
+/// executing user code.
+///
+/// Note that this does not actually implement `Iterator` as this struct needs
+/// to share access to the activation with you. We can't claim your activation
+/// and give it back in `next`, so we instead ask for it in `next`, which is
+/// incompatible with the trait.
+///
+/// This technically works with Array-shaped, non-Array objects, since we
+/// access arrays in this iterator the same way user code would. If it is
+/// necessary to only work with Arrays, you must first check for array storage
+/// before creating this iterator.
+///
+/// The primary purpose of `ArrayIterator` is to maintain lock safety in the
+/// presence of arbitrary user code. It is legal for, say, a method callback to
+/// mutate the array under iteration. Normally, holding an `Iterator` on the
+/// array while this happens would cause a panic; this code exists to prevent
+/// that.
+struct ArrayIterator<'gc> {
+    array_object: Object<'gc>,
+    index: u32,
+    length: u32,
+}
```

Comment:
Naming nit: Rust convention uses `Iter` for iterator types vs `Iterator` for traits, so this is better as `ArrayIter`.